c++/src/kj/compat/http.c++

```diff
@@ -2667,6 +2894,34 @@ private:
 
     Mask mask(maskKeyGenerator);
 
+    bool useCompression = false;
+    kj::Maybe<kj::Array<byte>> compressedMessage;
+    if (opcode == OPCODE_BINARY || opcode == OPCODE_TEXT) {
+      // We can only compress data frames.
+      KJ_IF_MAYBE(config, compressionConfig) {
+        KJ_DBG("[SEND] ORIGINAL size", message.size());
+        useCompression = true;
+        // Compress `message` according to `compressionConfig`s outbound parameters.
+        auto& compressor = KJ_ASSERT_NONNULL(compressionContext);
+        if (config->outboundNoContextTakeover) {
+          // We must reset context on each message.
+          compressor.reset();
+        }
+        KJ_SWITCH_ONEOF(compressor.processMessage(message)) {
+          KJ_CASE_ONEOF(e, kj::Exception) {
+            // TODO(now): What should happen here? We failed to compress...
+            KJ_DBG(e);
+          }
+          KJ_CASE_ONEOF(compressed, kj::Vector<kj::byte>) {
+            KJ_DBG("[SEND] Compressed size", compressed.size());
+            auto& innerMessage = compressedMessage.emplace(compressed.releaseAsArray());
+            message = innerMessage.slice(0, innerMessage.size() - 4);
+            // Strip 0x00 0x00 0xFF 0xFF off the tail.
```

Comment:
Should we assert that the buffer actually ends with that?

Once #1504 is in, you should even be able to write this as:

`KJ_ASSERT(innerMessage.endsWith({0x00, 0x00, 0xFF, 0xFF}));`

Reply:
https://github.com/capnproto/capnproto/pull/1505/commits/b2d1fada2df91e6d925096b79a4df2b627867bdc

Reply:
Followed up on this now that #1504 was merged: https://github.com/capnproto/capnproto/pull/1505/commits/4b3520dc75dc38872ec187f5734895d614beb254